engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java

```diff
@@ -0,0 +1,230 @@
+import org.springframework.stereotype.Service;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+@Service
+public class DockerService extends RService {
+    @Autowired
+    @Qualifier("analysisTaskExecutor")
+    private ThreadPoolTaskExecutor executor;
+    @Autowired
+    private DockerProperties dockerProperties;
+
+    @Override
+    public Overseer analyze(
+            AnalysisSyncRequestDTO analysis, File file, DescriptorBundle bundle,
+            KrbConfig krbConfig, BiConsumer<String, String> callback, Integer updateInterval) {
+        Instant started = Instant.now();
+        StringBuffer stdout = new StringBuffer();
+        BiConsumer<String, String> callbackWithStdOut = callback.andThen((stage, out) -> {
+            stdout.append(out);
+        });
+
+        CompletableFuture<ExecutionOutcome> future = CompletableFuture.supplyAsync(() -> {
+            try (DockerClient client = dockerClient()) {
+                String image = analysis.getDockerImage();
+                pullImage(image);
+                log.info("Creating container for image: {} and analysis Id: {}", image, analysis.getId());
+                CreateContainerResponse container = client.createContainerCmd(image)
+                        .withStdinOpen(true)
+                        .withTty(true)
+                        .exec();
+                log.info("Container created for image {} with container id {}", image, container.getId());
+                String containerId = container.getId();
+                runContainer(client, containerId, file, analysis, callbackWithStdOut);
+                return new ExecutionOutcome(Stage.COMPLETED, null, "");
+            } catch (IOException e) {
+                log.error("Execution [{}] failed:", analysis.getId(), e);
+                callbackWithStdOut.accept(Stage.EXECUTE, e.getMessage());
+                return new ExecutionOutcome(Stage.COMPLETED, e.getMessage(), "");
+            }
+        }, executor);
+        return new SQLService.SqlOverseer(analysis.getId(), started, stdout, future);
+    }
+
+    private void runContainer(DockerClient dockerClient, String containerId, File analysisDir, AnalysisSyncRequestDTO analysis, BiConsumer<String, String> callback) {
+        String execFileName = analysis.getExecutableFileName();
+
+        log.info("Starting container: {}", containerId);
+        dockerClient.startContainerCmd(containerId).exec();
+
+        String codeDir = "/bin";
+        log.info("Copying study into container: {}", containerId);
+        dockerClient.copyArchiveToContainerCmd(containerId)
+                .withHostResource(analysisDir.getPath())
+                .withRemotePath(codeDir).exec();
+
+        String baseDir = codeDir + "/" + analysisDir.getName() + "/";
+        String workingDir = baseDir + (execFileName.split("/").length > 1 ? execFileName.split("/")[0] : "");
```

Comment:
I don't understand the intent of string manipulation here, could you please clarify it?

Reply:
Sure, this code dynamically constructs the directory path inside the Docker container where the R script is located and will be executed. Also consider the possibility of the script being in a subdirectory within the analysis directory.